test/new-e2e/pkg/e2e/suite.go

```diff
@@ -496,6 +497,20 @@ func (bs *BaseSuite[Env]) TearDownSuite() {
 	if bs.params.devMode {
 		return
 	}
+	// If a Kubernetes environment is detected, we dump the cluster state when tearing down the suite
+	//TODO: *environments.Kubernetes string should be replaces by reflect.TypeOf(&environments.Kubernetes{}) once we move types.go to its own package to avoid import cycles
+	if fmt.Sprintf("%v", reflect.TypeOf(bs.env)) == "*environments.Kubernetes" {
```

Comment:
<!-- Questions are appropriate if you have a potential concern but are not quite sure if it's relevant or not. Asking the author for clarification or investigation can lead to a quick resolution. -->
**❓ question**
What if instead we had an `env.OnTearDown` callback, added to an environment interface? We could also do other clean up operations, thinking calling flush on fakeintake when in dev mode.

Reply:
Agree, as discussed offline. I will create a `TearDown` interface and check if the environment implements it

Reply:
I agree that we need to know how the cluster was provisioned to dump its state. This is not something `Environment` offer because it does not need to know how it has been provisionned. 
To have something that does not depend on the provisionning method I used a `DumpKubernetesClusterState` function that is supposed to "detect" the underlying provisionning method. But it creates ugly code and would become even worse when we add other Cloud Providers...
I am not sure to see how we would leverage provisionners to achieve that, because for now provisionners are only executing code when we create stack, while we'd need something that dump cluster state when the Suite is teared down

Reply:
The provisioner interface is not set in stone. For instance we could add an optional `Diagnose()` interface that we would automatically call on provisioning failure.

Reply:
Why not environments? Environments are a bunch of clients, and I can call my clients to do their cleanup. I don't care if the client was provisioned through EKS or kind, I just want to inform it that the test is ending, or starting, so that it does actions if it is interested on it.

Example, what if we informed fakeintake that the test is starting, in dev mode, so that it automatically flushes the server?

Reply:
I think that we could have two different things. I agree with Vincent that the Kubernetes cluster dump really needs the info about how the Cluster was provisioned. So it is hard to put that in the Environment, otherwise you need your cluster dump function to "detect" the way the cluster was created and it feels very hacky.

However I think your point is relevant and we could definitely have some environment that implements a `TearDown()` interface that would be called on the `Suite` TearDown(). It totally makes sense for flushing the fakeintake for example since you do not care how the fakeintake was provisionned untill you have the client to interact with it.

Small note: If we want the TearDown to work even when Pulumi failed to provision the stack it is harder because we will not be able to use the Fakeintake client directly